sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosDiagnosticsContext.java

```diff
@@ -560,6 +585,249 @@ public String toJson() {
         }
     }
 
+    /**
+     * Gets the UserAgent header value used by the client issueing this operation
+     * NOTE: this information is not included in the json representation returned from {@link #toJson()} because it
+     * is usually only relevant when thresholds are violated, in which case the entire diagnostics json-string is
+     * included. Calling this method will lazily collect the user agent - which can be useful when writing
+     * a custom {@link CosmosDiagnosticsHandler}
+     * @return the UserAgent header value used for the client that issued this operation
+     */
+    public String getUserAgent() {
+        if (this.diagnostics == null) {
+            return "";
+        }
+
+        CosmosDiagnostics diagnostics = this.diagnostics.peekFirst();
+        if (diagnostics == null) {
+            return "";
+        }
+
+        return diagnostics.getUserAgent();
+    }
+
+    /**
+     * Returns the set of contacted regions
+     * NOTE: this information is not included in the json representation returned from {@link #toJson()} because it
+     * is usually only relevant when thresholds are violated, in which case the entire diagnostics json-string is
+     * included. Calling this method will lazily collect the user agent - which can be useful when writing
+     * a custom {@link CosmosDiagnosticsHandler}
+     * @return the set of contacted regions
+     */
+    public String getConnectionMode() {
+        if (this.diagnostics == null) {
```

Comment:
for connection mode, why do we need to go through all diagnostics and iterator over them? Can't we use the connection mode set by customer on the client?

Reply:
no easy way to access the client here.

Reply:
And i don't want to add new string field just for the rare case where this API is called.

Reply:
To be honest, passing the connection mode and user agent through DiagnosticsProvider seems to be easier and more efficient option, given its just 2 strings, as compared to lazily calculating this even if it is not getting called every time. Also, WMT might use this API a lot, so that would save overhead for them as well. Just my opinion.